### ISSUE TITLE
Speed ups

### ISSUE THREAD

**Issue by AnthonyLim23** [ISISMuon/MuonDataLib#18]
We could look at speed ups. The following spec was used:

there are 10001 frames, each frame consists of 64 channels (channel = detector), and each channel trace consists of 30000 samples (sample time is 1sample = 1ns).
Each trace has between 0 and 200 (inclusive) simulated pulses.
-	Pulses are triangular in shape, with start time exponentially generated with mean lifetime of 2200ns.
-	Pulse height is uniformly random between 30 and 70 units (voltage units are arbitrary).
-	Pulse width is uniformly random between 20ns and 50ns.
-	The peak of the triangle is uniformly random to occur anywhere between the start and the end.
-	Event formation uses a fixed threshold discriminator with a threshold of 10 units.

and got times of:

about 4 sec 
about 3.28 sec (for large chunks)

**Comment by AnthonyLim23** [comment]
@modularius is helping me get setup so I can produce my own example files. 

I am going to edit the code to store the time for using h5py, the total number of frames and total number of counts. This will create some slow down, but it will not be significant.

**Comment by AnthonyLim23** [comment]
I needed to install scipy and I added matplotlib (as I know the syntax) to a test environment. The following code will produce data and do fits (linear, log(N) and Nlog(N))

```
from MuonDataLib.plot.basic import Figure
from MuonDataLib.data.loader.load_event import LoadEventData
import os
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit


def linear(x, a, b):
    return a*x + b

def log(x, a, b):
    return a*np.log(x) + b

def NlogN(x, a, b):
    return a*x*np.log(x) + b


files = ['LargeSimulatedTest.nxs',
         'LargeSimulatedTestHighChunk.nxs']


events = []
frames = []
h5py = []
tot = []
load = []

for file_name in files:

    file = os.path.join(os.path.dirname(__file__), file_name)

    # create data loaders and get histograms

    data_p = LoadEventData()
    data_p.load_data(file)

    h5py.append(data_p.h5py_time)
    tot.append(data_p.total_time)
    events.append(data_p.N_events)
    frames.append(data_p.frames)
    load.append(data_p.load_time)

h5py = np.asarray(h5py)
tot = np.asarray(tot)
events = np.asarray(events)
frames = np.asarray(frames)
load = np.asarray(load)


plt.plot(events, load, 'kx', label='data')

params, _ = curve_fit(linear, events, load, [1, 0])
plt.plot(events, linear(events, *params), 'r--', label='linear fit')

params, _ = curve_fit(log, events, load, [1, 0])
plt.plot(events, log(events, *params), 'b--', label='log fit')


params, _ = curve_fit(NlogN, events, load, [1, 0])
plt.plot(events, NlogN(events, *params), 'g--', label='NlogN fit')

plt.legend()

plt.show()

```

**Comment by AnthonyLim23** [comment]
I have received some files with 2000 frames for 64, 128, 200, 400 and 800 detectors. However, the time scaling in my code looks to be poor (I think quadratic)

![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/f14af60b-05f8-4c75-acef-b7b2e6b5c449)

I have checked and the h5py part is linear scaling.

**Comment by AnthonyLim23** [comment]
the lines are linear, NlogN and logN fits

**Comment by AnthonyLim23** [comment]
I have looked into this and my current approach is quadratic scaling with the number of detectors. Lets assume this is all for one frame and one period, for simplicity. If i have N detectors and each detector see M events then the total length of my arrays is NM. I have arrays for the event times, event amplitudes and detector ID's. So for each detector I want to check the array for which events are relevant. Hence, we have N*NM look ups and N^2 scaling. 

I will have a look and see if I can find anything better.

**Comment by AnthonyLim23** [comment]
I have added cython, which is a mixed language. So would require some changes to the code base to move from a pure python package (and has some other consequences). I have also made a second plot (the data points are the same for both) and the new plot has a solid red line at 0.5 million events per second. 

![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/d61883f8-011e-4bbb-99f1-34634101da13)

In all cases we are above the worst case. The np.where does much better for low number of events, but I think it will stay above the line for larger values as the trend seems to be a decay that is levelling off. The other methods are all straight lines (linear scaling) and are essentially doing the same thing, just with some tricks added in.

**Comment by AnthonyLim23** [comment]
![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/0088a3f4-b06d-4bbc-ab2a-d2748e7304df)

so cppyy lets me write c++ into python. It has some amazing times (its slowest was 2.8 sec for 800 detectors, whereas basic cython with 64 detectors was 2.3 seconds). However, the numbers for cppyy do seem a bit erratic 
0.25, 0.17, 0.3, 0.85, 2.8 for 64, 128, 200, 400, 800 detectors respectively

**Comment by wlav** [comment]
> However, the numbers for cppyy do seem a bit erratic

How do you time this? The first call into any JITed function includes compilation of a wrapper, which takes time and in particular may incur more serialization or loading of headers, ie. disk access. Maybe the erratic timing clears out if you lift 1 call out of the timed loop?

> numba uses the numba library, it compiles the code at runtime (so should be faster)

Note that there's now some basic Numba support for cppyy as well. We have some students working on Numba this Summer, so if you have an actual, real world, use case that can run relatively easily standalone, we'd be very interested.

**Comment by AnthonyLim23** [comment]
I have written some test code that handles things differently. Now the loading does scale linearly with number of detectors and number of frames. However, the histograms are very slow. I have attached some plots (they show similar things for frames/N). 

![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/35f747d7-f55f-40d6-b6a0-dc94c6b3504c)

The other downside is that the total time (load + histogram) is below the 0.5 million events per second. 

I will double check the scaling for the histograms in the current version.

**Comment by AnthonyLim23** [comment]
All I have done is move the quadratic behaviour between the 2 versions of the code. The test version is linear in loading and the current version is linear in histograms 
![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/caf3f66a-bae1-4722-abe2-af62276e6246)

I will have a look to see if I can get the test version to be linear in histograms too

**Comment by AnthonyLim23** [comment]
looking at the code, the difference is purely where we match the event to the detector. As part of load or the histogram-ing. That seems to be what causes the non-linear growth.

**Comment by AnthonyLim23** [comment]
I have found a method that might work. It has a single cython loop (removing it does make the code slower, but the scaling is the same).

I have linear scaling for both histograms and load as a function of the number of detectors
![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/4ab414f6-3325-4973-b013-ba487b268191)
It is also far above our minimum of 0.5e6 events per second. 

I also have linear scaling for both histograms and load as a function of number of frames (64 detectors). The worst case data point has 8.8735e6 events per second, so about a factor 17 better than the minimum. 

![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/00f1f153-c060-4b21-9705-d500beb10714)

However, I still need to investigate the following:

- [ ] Being able to isolate frames for filtering
- [ ] Scaling of filtering 
- [ ] Multiple periods (not got data for this, but I can fake it by assigning periods based on if its a even or odd frame)

At present I have N lists of events (one for each detector). However, I need to be able to identify which events belong to each frame for filtering and multiple period data. So I need to make a N*F lists, where F is the number of frames. This will allow me to say which events occurred on a specific detector in a given frame.

**Comment by AnthonyLim23** [comment]
I just did it for a pure python implementation.

For the number of detectors: 
![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/817b828b-0f38-428e-b646-52459048b1e9)

A quick and dirty estimate gives a predicted time of 400 seconds for 1000 detectors (point at 43 sec for 128 det, 318 sec for 800 det). Then as as estimate 1000 detectors is about 2e8 events (0.5-ish for 200 and 1.5-ish for 800). So it is about 5e5 events per second, so right on the acceptable line. 

everything looks good, just the performance gets close to the limit.

For number of frames:

![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/f578c5c9-6627-47cf-99e9-83efa58c30ac)

**Comment by AnthonyLim23** [comment]
My latest test allows me to identify specific events with a detector and a frame. The detector's are needed for histograms and the frames for filtering. It does show linear scaling for both load and histograms (assuming all frames for now). However, this pure python implementation is slow and below our base line. 

![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/f4b27952-b889-405f-8cdc-657d021e277f)

**Comment by AnthonyLim23** [comment]
With a bit of cython I was able to get linear scaling for both number of detectors and number of frames. For number of detectors:
![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/f978bc4a-b3db-4010-a1f1-cf790b7233df)
The rate seems to be about 7e6, so 14 times better than the base line. 

For the number of frames I have

![image](https://github.com/ISISMuon/MuonDataLib/assets/25006392/f58531f8-8dba-4897-ae77-8d54c4a6229e)

so again linear scaling and about 7e6 events per second. 

All I have done is put the 2 main loops into cython. I have not done any filtering and I have not done anything for periods. 

I have assumed that all frames are the same length.

**Comment by AnthonyLim23** [comment]
after a bit more investigating. It looks like my cython code is faster than c++ code via cython. I assume they use some extra cleaver efficiency that I am not.

**Comment by AnthonyLim23** [comment]
The latest version is designed to allow for filtering of partial frames. 

However, Peter has said that if part of a frame is filtered out then the whole frame should be discarded. This could allow for some improvements. I will have a look into it.

**Comment by AnthonyLim23** [comment]
I have been able to simplify my test code with the following assumptions:

- only include whole frames
- can discard any events after some upper limit (in this case 30)
- single period data only
- constant width bins (0.5 in my example)
- no multi-processing 

 Instead of creating a matrix of events and then constructing the histograms, I construct the histograms directly (and manually). The run times are significantly better and give a ball park for what the final version could achieve. As this is missing

- periods
- filtering

To test these I have ran each file 10 times, then used the mean value for the plot and the standard deviation for the error bars. 
I have done 2000 frames for different numbers of detectors (N), namely 128, 200, 400, 800.
I have also done different numbers of frames for 64 detectors, namely 4000, 7000, 10000, 12000, 16000
On my laptop I got:
![time_frames](https://github.com/ISISMuon/MuonDataLib/assets/25006392/a2419602-0250-432c-a31e-ddc1d6739fa1)
![time_N](https://github.com/ISISMuon/MuonDataLib/assets/25006392/64939764-fdb3-4463-9d59-0779960f4c3e)
so in both cases a nice linear scaling.

In terms of rates we get
![rate_frames](https://github.com/ISISMuon/MuonDataLib/assets/25006392/76e42d06-fffb-4fba-a99f-dec43c974b3d)
![rate_N](https://github.com/ISISMuon/MuonDataLib/assets/25006392/ccc8ec1e-dba0-4af6-953a-c31d0e47afaa)

so for the number of frames the rate is always better than 3e7, so about 60 times better than the baseline (5e5).
For the different number of detectors it was always better than 1.5e8, so about 300 times better.
